File: sqb-api-skills/sqb-query/reference/QueryExample.java
package com.example.shouqianba;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

/**
 * 收钱吧订单查询示例
 *
 * 支持单次查询和轮询查询两种模式
 */
public class QueryExample {

    private static final String API_BASE = "https://vsi-api.shouqianba.com";
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");
    private static final OkHttpClient client = new OkHttpClient();
    private static final ObjectMapper mapper = new ObjectMapper();

    // 最终状态集合
    private static final Set<String> FINAL_STATUSES = new HashSet<>(Arrays.asList(
        "PAID", "PAY_CANCELED", "REFUNDED", "PARTIAL_REFUNDED"
    ));

    private String terminalSn;
    private String terminalKey;

    public QueryExample(String terminalSn, String terminalKey) {
        this.terminalSn = terminalSn;
        this.terminalKey = terminalKey;
    }

    /**
     * 查询订单（通过商户订单号）
     */
    public QueryResult queryByClientSn(String clientSn) throws IOException {
        ObjectNode body = mapper.createObjectNode();
        body.put("terminal_sn", terminalSn);
        body.put("client_sn", clientSn);
        return doQuery(body);
    }

    /**
     * 查询订单（通过收钱吧订单号）
     */
    public QueryResult queryBySn(String sn) throws IOException {
        ObjectNode body = mapper.createObjectNode();
        body.put("terminal_sn", terminalSn);
        body.put("sn", sn);
        return doQuery(body);
    }

    /**
     * 轮询查询（等待最终状态）
     *
     * @param clientSn       商户订单号
     * @param maxWaitSeconds 最大等待时间
     * @return 最终查询结果
     */
    public QueryResult pollUntilFinal(String clientSn, int maxWaitSeconds)
            throws IOException, InterruptedException {

        long startTime = System.currentTimeMillis();

        while (true) {
            long elapsed = (System.currentTimeMillis() - startTime) / 1000;

            if (elapsed > maxWaitSeconds) {
                return new QueryResult("TIMEOUT", null, "轮询超时");
            }

            QueryResult result = queryByClientSn(clientSn);
            if (result.isFinal()) {
                return result;
            }

            System.out.println("[" + elapsed + "s] 状态: " + result.orderStatus + "，继续轮询...");

            // 轮询间隔：前60秒3秒，之后10秒
            int interval = elapsed < 60 ? 3 : 10;
            Thread.sleep(interval * 1000L);
        }
    }

    private QueryResult doQuery(ObjectNode body) throws IOException {
        String bodyStr = mapper.writeValueAsString(body);
        String sign = md5(bodyStr + terminalKey);

        Request request = new Request.Builder()
            .url(API_BASE + "/upay/v2/query")
            .addHeader("Authorization", terminalSn + " " + sign)
            .addHeader("Content-Type", "application/json; charset=utf-8")
            .post(RequestBody.create(bodyStr, JSON))
            .build();

        try (Response response = client.newCall(request).execute()) {
            String respBody = response.body().string();
            JsonNode resp = mapper.readTree(respBody);

            if (!"200".equals(resp.path("result_code").asText())) {
                return new QueryResult("ERROR", null,
                    resp.path("error_message").asText("通信失败"));
            }

            JsonNode bizResponse = resp.path("biz_response");
            if (!"SUCCESS".equals(bizResponse.path("result_code").asText())) {
                return new QueryResult("ERROR", null,
                    "查询失败: " + bizResponse.path("result_code").asText());
            }

            JsonNode data = bizResponse.path("data");
            return new QueryResult(
                data.path("order_status").asText(),
                data,
                "查询成功"
            );
        }
    }

    private static String md5(String input) {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] digest = md.digest(input.getBytes(StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder();
            for (byte b : digest) {
                sb.append(String.format("%02x", b));
            }
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * 查询结果
     */
    public static class QueryResult {
        public final String orderStatus;
        public final JsonNode data;
        public final String message;

        public QueryResult(String orderStatus, JsonNode data, String message) {
            this.orderStatus = orderStatus;
            this.data = data;
            this.message = message;
        }

        public boolean isFinal() {
            return FINAL_STATUSES.contains(orderStatus);
        }

        @Override
        public String toString() {
            return "QueryResult{orderStatus='" + orderStatus + "', message='" + message + "'}";
        }
    }
}

File: sqb-api-skills/sqb-query/reference/query_example.py
"""
收钱吧订单查询示例 (Python)

支持单次查询和轮询查询两种模式
"""

import hashlib
import json
import time

import requests

API_BASE = "https://vsi-api.shouqianba.com"

# 最终状态集合
FINAL_STATUSES = {"PAID", "PAY_CANCELED", "REFUNDED", "PARTIAL_REFUNDED"}


def md5_sign(body_str: str, key: str) -> str:
    """计算 MD5 签名"""
    content = body_str + key
    return hashlib.md5(content.encode("utf-8")).hexdigest()


class ShouqianbaQueryClient:
    """收钱吧查询客户端"""

    def __init__(self, terminal_sn: str, terminal_key: str):
        self.terminal_sn = terminal_sn
        self.terminal_key = terminal_key

    def _request(self, body: dict) -> dict:
        """发送查询请求"""
        body_str = json.dumps(body, ensure_ascii=False)
        sign = md5_sign(body_str, self.terminal_key)

        headers = {
            "Authorization": f"{self.terminal_sn} {sign}",
            "Content-Type": "application/json; charset=utf-8",
        }

        resp = requests.post(
            f"{API_BASE}/upay/v2/query",
            data=body_str.encode("utf-8"),
            headers=headers,
            timeout=30,
        )
        return resp.json()

    def query(self, client_sn: str = None, sn: str = None) -> dict:
        """
        查询订单

        Args:
            client_sn: 商户订单号（与 sn 二选一）
            sn: 收钱吧订单号（优先级高于 client_sn）

        Returns:
            {"order_status": ..., "data": ..., "is_final": bool}
        """
        body = {"terminal_sn": self.terminal_sn}
        if sn:
            body["sn"] = sn
        elif client_sn:
            body["client_sn"] = client_sn
        else:
            raise ValueError("sn 和 client_sn 至少传一个")

        result = self._request(body)

        if result.get("result_code") != "200":
            return {
                "order_status": "ERROR",
                "data": None,
                "is_final": False,
                "message": result.get("error_message", "通信失败"),
            }

        biz = result.get("biz_response", {})
        if biz.get("result_code") != "SUCCESS":
            return {
                "order_status": "ERROR",
                "data": None,
                "is_final": False,
                "message": f"查询失败: {biz.get('result_code')}",
            }

        data = biz.get("data", {})
        order_status = data.get("order_status", "")

        return {
            "order_status": order_status,
            "data": data,
            "is_final": order_status in FINAL_STATUSES,
            "message": "查询成功",
        }

    def poll_until_final(self, client_sn: str, max_wait_seconds: int = 120) -> dict:
        """
        轮询查询，等待最终状态

        Args:
            client_sn: 商户订单号
            max_wait_seconds: 最大等待时间（秒）

        Returns:
            最终查询结果
        """
        start_time = time.time()

        while True:
            elapsed = time.time() - start_time

            if elapsed > max_wait_seconds:
                return {
                    "order_status": "TIMEOUT",
                    "data": {"client_sn": client_sn},
                    "is_final": False,
                    "message": f"轮询超时（{max_wait_seconds}秒）",
                }

            result = self.query(client_sn=client_sn)

            if result["is_final"]:
                return result

            print(f"[{elapsed:.0f}s] 状态: {result['order_status']}，继续轮询...")

            # 轮询间隔：前60秒3秒，之后10秒
            interval = 3 if elapsed < 60 else 10
            time.sleep(interval)


if __name__ == "__main__":
    client = ShouqianbaQueryClient(
        terminal_sn="your_terminal_sn",
        terminal_key="your_terminal_key",
    )

    # 单次查询
    result = client.query(client_sn="20230615143052001")
    print(f"查询结果: {result}")

    # 轮询查询（等待支付完成）
    # result = client.poll_until_final("20230615143052001", max_wait_seconds=120)
    # print(f"最终结果: {result}")
